### ISSUE TITLE
Backward call per loss

### ISSUE THREAD

**Issue by MrNeRF** [MrNeRF/gaussian-splatting-cuda#147]
Try to compute the backward after every loss. Run the benchmark. Verify memory consumption.

https://github.com/MrNeRF/gaussian-splatting-cuda/blob/9bfd4f22bf191744e9f03df5d1b59bd500a3034e/src/trainer.cpp#L220

![Image](https://github.com/user-attachments/assets/d3b440c8-6467-4df9-bbfd-9490733709e6)

Inspired by: https://x.com/gabriberton/status/1796531585958985941?s=61

**Comment by MrNeRF** [comment]
Probably only works by separating the photometric losses from the regularization, i.e. in code:
loss = ... 
loss.backward();
opacity_reg_loss.backward();
scale_reg_losd.backward();

**Comment by dakshshah03** [comment]
Is the idea to split the compute_loss function into separate TV, SSIM, and L1 functions, then perform backprop on each one separately?

**Comment by MrNeRF** [comment]
Yeah. But I think you will need to group the photometric losses, i.e in pseudo code:

Loss = l1(img,gt) + ssim(img,gt)
Loss.backward
...

In our case there would be 3 backward calls in total.

I wonder if this has a measureable impact on memory or runtime. Does it break, when you don't group the photometric losses but call instead backward 4 times?

**Comment by Tomoya-Matsubara** [comment]
I did a quick experiment in my environment, and the result looks like this:

<img src="https://github.com/user-attachments/assets/c3020484-1050-4853-bc0b-69bcb551093a" width="500" />

Although it might reduce memory usage, I'm not sure if it's always the case. This is a result of a different run:

<img src="https://github.com/user-attachments/assets/e5160ddc-3874-471e-80f1-072276a26237" width="500" />

**Comment by MrNeRF** [comment]
From the figures it seems on average to have lower mem consumption but not on peak. Interesting.

Which scene is this? And I assume you have the 3 backward calls? @Tomoya-Matsubara

**Comment by Tomoya-Matsubara** [comment]
> Which scene is this? And I assume you have the 3 backward calls?

It's drjohnson from [tandt_db](https://repo-sam.inria.fr/fungraph/3d-gaussian-splatting/datasets/input/tandt_db.zip).

There were 4 backward calls,

1. Photometric loss
2. Scale regularization loss
3. Opacity regularization loss
4. TV loss

but since I didn't enable the bilateral grid, it was 3 in reality.

**Comment by MrNeRF** [comment]
Thank you! 

Could you also do a run with bilateral grid enabled to see the impact? 

Then, I think you can create a pr to get this into master.

**Comment by Tomoya-Matsubara** [comment]
Sure!

**Comment by MrNeRF** [comment]
Btw, if someone’s lurking here, there’s another good issue to get a bit more familiar with the code open to be grabbed:
https://github.com/MrNeRF/gaussian-splatting-cuda/issues/152
